File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = []//init a stack
    for(let c of s){
        if(c=='(' || c == '[' || c=='{'){
            stack.push(c)//push all the opening parenthesis 
        }
        else{
            if(!stack.length || 
             (c == ')' && stack[stack.length-1]!=='(') ||
             (c == ']' && stack[stack.length-1]!=='[') ||
             (c == '}' && stack[stack.length-1]!=='{') )
                {
                return false;
        
        }
        
        stack.pop()    
        }
        
    }
    return !stack.length;
}

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
    stack<char> st;//creating a stack of char
    for(char c :s)
        {
            if(c =='(' || c =='[' || c =='{'){
                st.push(c);// if there is any opening parenthesis add into stack
            }
            else{
                if(st.empty() || //checking whether the incoming bracket is closed and 
                ( c == ')' && st.top()!= '(') || // if so checking whether the char in the top of the stack is matching
                ( c == '}' && st.top() != '{') ||
                ( c == ']' && st.top() != '[') )
                {
                return false;}//if any of the condition fails return false 
                
        st.pop();//pop every char inside the for loop 
        }
        
    
        }
    return st.empty();// after poping the stack must be empty 
    }
};